sycl/tools/sycl-sanitize/collector.cpp

```diff
@@ -0,0 +1,216 @@
+                                 pi_uint32) {
+  AllocationInfo Info;
+  Info.Location = GS->LastTracepoint;
+  Info.Length = Size;
+  Info.Kind = AllocKind::shared;
+  GS->ActivePointers[*ResultPtr] = Info;
+}
+
+static void handleUSMFree(const pi_plugin &, std::optional<pi_result>,
+                          pi_context, void *Ptr) {
+  if (GS->ActivePointers.count(Ptr) == 0) {
+    std::cerr << "Attempt to free pointer " << std::hex << Ptr;
+    std::cerr << " that was not allocated with SYCL USM APIs.\n";
+    std::cerr << "  Location: function " << GS->LastTracepoint.Function;
+    std::cerr << " at " << GS->LastTracepoint.Source << ":";
+    std::cerr << std::dec << GS->LastTracepoint.Line << "\n";
+    std::terminate();
+  }
+  GS->ActivePointers.erase(Ptr);
+}
+
+static void handleMemBufferCreate(const pi_plugin &, std::optional<pi_result>,
+                                  pi_context, pi_mem_flags, size_t Size,
+                                  void *HostPtr, pi_mem *,
+                                  const pi_mem_properties *) {
+  for (const auto &Alloc : GS->ActivePointers) {
+    const void *Begin = Alloc.first;
+    const void *End =
+        static_cast<const char *>(Alloc.first) + Alloc.second.Length;
+    // Host pointer was allocated with USM APIs
+    if (HostPtr >= Begin && HostPtr <= End) {
+      bool NeedsTerminate = false;
+      if (Alloc.second.Kind != AllocKind::host) {
+        std::cerr << "Attempt to construct a buffer with non-host pointer.\n";
+        NeedsTerminate = true;
+      }
+
+      const void *HostEnd = static_cast<char *>(HostPtr) + Size;
+      if (HostEnd > End) {
+        std::cerr << "Buffer size exceeds allocated host memory size.\n";
+        NeedsTerminate = true;
+      }
+
+      if (NeedsTerminate) {
+        std::cerr << "  Allocation location: ";
+        std::cerr << " function " << Alloc.second.Location.Function << " at ";
+        std::cerr << Alloc.second.Location.Source << ":"
+                  << Alloc.second.Location.Line << "\n";
+        std::cerr << "  Buffer location: ";
+        std::cerr << " function " << GS->LastTracepoint.Function << " at ";
+        std::cerr << GS->LastTracepoint.Source << ":" << GS->LastTracepoint.Line
+                  << "\n";
+        std::terminate();
+      }
+      break;
+    }
+  }
+}
+
+XPTI_CALLBACK_API void tpCallback(uint16_t trace_type,
+                                  xpti::trace_event_data_t *parent,
+                                  xpti::trace_event_data_t *event,
+                                  uint64_t instance, const void *user_data);
+
+XPTI_CALLBACK_API void xptiTraceInit(unsigned int /*major_version*/,
+                                     unsigned int /*minor_version*/,
+                                     const char * /*version_str*/,
+                                     const char *StreamName) {
+  if (std::string_view(StreamName) == "sycl.pi.debug") {
+    GS = new GlobalState;
```

Comment:
Do you really need some raw memory?

Reply:
Yes. Some SYCL users use SYCL APIs in global scope, and to work correctly, `GlobalState` has to outlive everything else in the application. Pointer is trivially destructible, so it'll be alive as long as process is alive. And it'll allow us to capture all the calls.